crates/engine/tree/src/tree/root.rs

```diff
@@ -106,86 +221,673 @@ where
+                self.calculating_root = false;
+
+                let has_new_proofs = !current_multiproof.account_subtree.is_empty() ||
+                    !current_multiproof.storages.is_empty();
+                let all_proofs_received = stats.proofs_processed >= stats.updates_received;
+                let no_pending = !self.proof_sequencer.has_pending();
+
+                if has_new_proofs {
+                    trace!(
+                        target: "engine::root",
+                        account_proofs = current_multiproof.account_subtree.len(),
+                        storage_proofs = current_multiproof.storages.len(),
+                        "Spawning subsequent root calculation"
+                    );
+                    self.spawn_root_calculation(std::mem::take(current_multiproof));
+                    None
+                } else if all_proofs_received && no_pending {
+                    debug!(
+                        target: "engine::root",
+                        total_updates = stats.updates_received,
+                        total_proofs = stats.proofs_processed,
+                        roots_calculated = stats.roots_calculated,
+                        "All proofs processed, ending calculation"
+                    );
+                    Some(Ok((*current_root, trie_updates.clone())))
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /// Handle internal message channel errors
+    fn handle_internal_error() -> StateRootResult {
+        error!(target: "engine::root", "Internal message channel closed unexpectedly");
+        Err(ParallelStateRootError::Other("Internal message channel closed unexpectedly".into()))
+    }
+
+    fn run(mut self) -> StateRootResult {
+        let mut stats = TaskStats::default();
+        let mut current_multiproof = MultiProof::default();
+        let mut trie_updates = TrieUpdates::default();
+        let mut current_root = B256::default();
+
+        loop {
+            match self.state_stream.rx.try_recv() {
+                Ok(update) => {
+                    stats.updates_received += 1;
+                    trace!(
+                        target: "engine::root",
+                        len = update.len(),
+                        total_updates = stats.updates_received,
+                        "Received new state update"
+                    );
+                    Self::on_state_update(
+                        self.config.consistent_view.clone(),
+                        self.config.input.clone(),
+                        update,
+                        &mut self.state,
+                        self.proof_sequencer.next_sequence(),
+                        self.tx.clone(),
+                    );
+                }
+                Err(mpsc::TryRecvError::Empty) => {
+                    // No state updates available, try to process internal messages
+                    match self.rx.recv() {
```

Comment:
if we're unfortunate to not get any state updates before entering this branch, will this `recv` get stuck forever?